packages/network-controller/src/create-network-client-tests/rpc-endpoint-events.test.ts

```diff
@@ -0,0 +1,1290 @@
+            const rpcEndpointChainDegradedEventHandler = jest.fn();
+            messenger.subscribe(
+              'NetworkController:rpcEndpointChainDegraded',
+              rpcEndpointChainDegradedEventHandler,
+            );
+
+            await withNetworkClient(
+              {
+                providerType: networkClientType,
+                networkClientId: 'AAAA-AAAA-AAAA-AAAA',
+                messenger,
+                getRpcServiceOptions: () => ({
+                  fetch,
+                  btoa,
+                  policyOptions: {
+                    backoff: new ConstantBackoff(backoffDuration),
+                  },
+                }),
+              },
+              async ({ makeRpcCall, clock, chainId, rpcUrl }) => {
+                // The first time a block-cacheable request is made, the
+                // latest block number is retrieved through the block
+                // tracker first.
+                comms.mockRpcCall({
+                  request: {
+                    method: 'eth_blockNumber',
+                    params: [],
+                  },
+                  response: () => {
+                    clock.tick(DEFAULT_DEGRADED_THRESHOLD + 1);
+                    return {
+                      result: '0x1',
+                    };
+                  },
+                });
+                comms.mockRpcCall({
+                  request,
+                  response: () => {
+                    clock.tick(DEFAULT_DEGRADED_THRESHOLD + 1);
+                    return {
+                      result: 'ok',
+                    };
+                  },
+                  times: 2,
+                });
+
+                await makeRpcCall(request);
+                await makeRpcCall(request);
+
+                expect(
+                  rpcEndpointChainDegradedEventHandler,
+                ).toHaveBeenCalledTimes(1);
+                expect(
+                  rpcEndpointChainDegradedEventHandler,
+                ).toHaveBeenCalledWith({
+                  chainId,
+                  endpointUrl: rpcUrl,
+                  error: undefined,
+                  networkClientId: 'AAAA-AAAA-AAAA-AAAA',
+                  primaryEndpointUrl: rpcUrl,
+                });
+              },
+            );
+          },
+        );
+      });
+
+      it('does not publish the NetworkController:rpcEndpointChainDegraded event again if the max number of retries is reached in making requests to a failover endpoint', async () => {
+        const failoverEndpointUrl = 'https://failover.endpoint/';
+        const request = {
+          method: 'eth_gasPrice',
+          params: [],
+        };
+        const expectedError = createResourceUnavailableError(503);
+        const expectedDegradedError = new HttpError(503);
+
+        await withMockedCommunications(
+          { providerType: networkClientType },
+          async (primaryComms) => {
+            await withMockedCommunications(
```

Comment:
Hmm. Seeing 3 nested `with___` blocks to setup the environment is not ideal.

Not something to address in this PR, it's already large enough. But it would be great if we could pass in a list of mocked communication objects to `withNetworkClient` to make this more concise (and reduce indentation)